stablehlo/reference/NumPy.cpp

```diff
@@ -0,0 +1,323 @@
+  // alphabetical ordering relative to other header keys), however we cannot
+  // assume this to be unilaterally true. Therefore, apply a tight bound on the
+  // regex matching for dimension integrals.
+  std::regex dimRegex("[0-9]+");
+  std::smatch dimMatch;
+  std::string shapeString = header.substr(shapeOffset, shapeOffset - dimEnd);
+  std::vector<int64_t> shape(4);
+
+  while (std::regex_search(shapeString, dimMatch, dimRegex)) {
+    shape.push_back(std::stoi(dimMatch[0]));
+    shapeString = dimMatch.suffix();
+  }
+
+  return shape;
+}
+
+template <typename T>
+static llvm::Error readNumpyHeader(std::ifstream& in, size_t& wordSize,
+                                   ArrayRef<int64_t> shape) {
+  char magicString[kMagicStringSize];
+  if (!in.read(magicString, kMagicStringSize))
+    return llvm::createStringError(llvm::errc::io_error,
+                                   "Failed to read NumPy magic string.");
+
+  if (strncmp(magicString, kMagicString, kMagicStringSize))
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Invalid NumPy file format detected.");
+
+  // This utility is presently only used from the context of the check dialect
+  // for validation. It suffices to only support NumPy v1.0 for now.
+  char majorVersion, minorVersion;
+  in.read(&majorVersion, 1);
+  in.read(&minorVersion, 1);
+
+  if (majorVersion != kMajorVersion || minorVersion != kMinorVersion)
+    return llvm::createStringError(
+        llvm::errc::invalid_argument,
+        "Invalid NumPy version detected. Only v1.0 supported.");
+
+  char headerSizeBuffer[2];
+  if (!in.read(headerSizeBuffer, 2))
+    return llvm::createStringError(llvm::errc::io_error,
+                                   "Failed to read NumPy header size.");
+
+  const int headerSize = (headerSizeBuffer[0]) | (headerSizeBuffer[1] << 8);
+  std::string header(headerSize, '\0');
+  if (!in.read(header.data(), headerSize) || header.back() != '\n')
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Invalid NumPy header.");
+
+  header.erase(std::remove_if(header.begin(), header.end(), isspace),
+               header.end());
+
+  auto wordSizeOrError = parseDescrHeader<T>(header);
+  if (!wordSizeOrError)
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Failed to parse descr header.");
+
+  if (parseFortranOrderKey(header))
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Failed to parse fortran_order header.");
+
+  auto shapeOrError = parseShapeHeader(header);
+  if (!shapeOrError)
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Failed to parse shape header.");
+
+  shape = *shapeOrError;
+  wordSize = *wordSizeOrError;
+
+  return llvm::Error::success();
+}
+
+namespace {
+template <typename T>
+class ToNumpy {
```

Comment:
Let's block big endian machines in `To/FromNumpy` and give a precise error that we don't support it.

We already don't support reading big endian files. I'm hesitant to say "lets permit big endian to create files" when we don't have any tests. I.e. _I think_ MLIR dense elements store everything in little endian, and if we end up constructing out Tensor objects directly from those, and passing data to this function at a `char*` then its possible that everything will be little endian, and we'd be incorrectly marking it as BE.